File: 3867.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var gcdSum = function(nums) {
    const n = nums.length;
    if (n === 0) return 0;

    const prefixGcd = new Array(n);
    let maxSoFar = 0;

    for (let i = 0; i < n; i++) {
        maxSoFar = Math.max(maxSoFar, nums[i]);
        prefixGcd[i] = gcd(nums[i], maxSoFar);
    }

    prefixGcd.sort((a, b) => a - b);

    let totalSum = 0;
    for (let i = 0; i < Math.floor(n / 2); i++) {
        totalSum += gcd(prefixGcd[i], prefixGcd[n - 1 - i]);
    }

    return totalSum;
};

function gcd(a, b) {
    while (b !== 0) {
        const temp = b;
        b = a % b;
        a = temp;
    }
    return a;
}

File: 3867.py
from math import gcd
from typing import List


class Solution:
    def gcdSum(self, nums: List[int]) -> int:
        n = len(nums)
        if n == 0:
            return 0

        prefix_gcd = [0] * n
        max_so_far = 0

        for i, num in enumerate(nums):
            max_so_far = max(max_so_far, num)
            prefix_gcd[i] = gcd(num, max_so_far)

        prefix_gcd.sort()

        total_sum = 0
        for i in range(n // 2):
            total_sum += gcd(prefix_gcd[i], prefix_gcd[-1 - i])

        return total_sum
